### ISSUE TITLE
basic version of FOM for v2 - [closed]

### ISSUE THREAD

**Issue by NicolasGensollen** [aramis-lab/leaspy#301]
In GitLab by @caglayantuna on Dec 5, 2024, 17:58

_Merges fix_FOM -> v2_

### What does the code in the MR do ?

This MR adapts the `FitOuputManager` functions for v2. I made modifications in 2 files, `fit_output_manager.py` and `abstract_fit_algo.py` . Here are the main changes;

1. Convergence parameters were not being saved with v2. Before, two PDF files were being saved in v1 version, but this might be unnecessary. With this MR, the user can now save a single PDF file containing the plot convergence parameters in v2. I skipped some variables and used the `tracked_variables` in the `State` class. You will find an example below. We can discuss the content of the saved PDF and edit it accordingly.
2. Second issue to solve after this MR, the fit function was printing incorrect values at the end of last iteration. Now, by using ` loss = state.get_tensor_value("noise_std")` , fit function prints correct values. I think that `terminate_algo` function is not necessary anymore. One thing to note, documentation should be improved in api.py. In v2, we should use `obs_models` instead of `noise_models`. And, fit can print feature based noises by using `gaussian_diagonal` for example.
3. I removed the following lines from `abstract_fit_algo.py`

   ```python
                   if self.logs:
                       self.log_current_iteration(state)
   
   ```

   and added the following operation to `fit_output_manager`  to have all printing/saving tasks in the same place

   ```python
           model.state.save(
               self.path_save_model_parameters_convergence,
               iteration=iteration,
           )
   ```
4. Besides, several unused functions (such as related to `Realization`) and comments have been removed from `FitOutputManager` in order to clean it a little bit. For example, there is no `Plotter` in `FitOutputManager` anymore. In visualization folder, there 2 files `Plotter` and `Plotting`. We can discuss whether to merge them or remove one to reduce redundancy.
5. Both files were formatted with ruff.
6. There is one remaining issue which is `plot_patient_reconstructions` function. I removed it for now because I couldn't find how to do it without `compute_individual_tensorized` which doesn't exist anymore. I couldn’t find any equivalent in v2. I would be happy to hear your ideas about it. Then, I can add this function in this MR or we can do it later.

undefined

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Dec 9, 2024, 16:09

added 1 commit

<ul><li>7c42bb8d - fix_tests</li></ul>

[Compare with previous version](/icm-institute/aramislab/leaspy/-/merge_requests/179/diffs?diff_id=1205986551&start_sha=0ebe75f596fc5df46aa83f53f2eb2d7f402adb95)

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Dec 10, 2024, 09:48

requested review from @NicolasGensollen

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Dec 10, 2024, 10:05

I added a second commit to fix some of the failed tests caused by `fit_metrics` and `mixing matrix`. At the moment, there are 4 additional failing tests regarding to `noise_std` compared to v2 branch. In this MR, I used the `model.state.get_tensor_value("noise_std")`  to calculate the loss in `run_impl`, however, this term doesn't exist with some of the observational models such as bernoulli. This line works well with gaussian models. 

In v1, this loss was calculated [like this](https://gitlab.com/icm-institute/aramislab/leaspy/-/blob/master/leaspy/algo/fit/abstract_fit_algo.py?ref_type=heads#L129). Maybe we should add similar thing but I was not sure how to do it. It would be better to discuss together.

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Dec 10, 2024, 10:08

requested review from @JulietteOrtholand

**Comment by NicolasGensollen** [comment]
In GitLab by @JulietteOrtholand on Dec 17, 2024, 10:44

Thank you very much for this nice MR ! I have not yet look at the code but the description seems promising ! For your last point, I think the function you are looking for is now named `compute_individual_trajectory`.

**Comment by NicolasGensollen** [comment]
In GitLab by @JulietteOrtholand on Dec 20, 2024, 09:36

Regarding your question on what parameters to plot when the distribution is not gaussian. The observational models are using a `StatelessDistributionFamily` to make the link with the data. 

If we take the example of the Gaussian distribution, it uses `NormalFamily` that has an attribute `parameters` that are the two parameters that define the distribution (loc, scale). Then in the observational model `FullGaussianObservationModel` in `_gaussian.py`, the parameters of the distribution are linked with the variables defined in the DAG: `loc="model", scale="noise_std"`. We are not interested directly in `model` but rather in the nll associated to the observation model as discused. So here we want the scale, so we keep `noise_std`, and the nll.

Following the same idea for longitudinal data:
- **Bernouilli**: only one parameters `loc` that is the `model` variable of the DAG. So we keep only the nll,
- **Ordinal**: I guess is the same as we only have one parameter.

For survival data:
- **Weibull**: there are no direct link with the "model" itself. So here I would use the two parameters of the distribution `nu` and `rho`. Their interpretation is not straight foward but more information can be found [here](https://www.sciencedirect.com/science/article/pii/S095183201100175X) for example.


So to sum up I would say:

Quality check:
- nll in every case,

Parameters:
- Weibull: ($t_0$ + nu: XX ([-]), rho ([-])),
- Gaussian: (noise_std ([%] => multiply by 100))
- Ordinal: -
- Bernouilli: -

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Dec 20, 2024, 16:38

added 1 commit

<ul><li>cbf718c6 - adds plot patient rec and first attemp for generalized loss</li></ul>

[Compare with previous version](/icm-institute/aramislab/leaspy/-/merge_requests/179/diffs?diff_id=1218242350&start_sha=7c42bb8d5390834ab680a196a49e374080250b0f)

**Comment by NicolasGensollen** [comment]
In GitLab by @JulietteOrtholand on Jan 14, 2025, 13:31

A loss is associated to each observational model. If there is only one observational model in the model, then it is stored in nll_attach. But then, if there are more than one observational model, the nll associated to each of them is stored under a different name defined in `get_nll_attach_var_name` in `_base.py` of observational models. 

Thus what we want to print:
- if there is only one observational model: `nll_attached`
- if there is at least two observational model: the nll of each observational model and `nll_attached` (which is the sum of all of them)

In the case of joint model (2 observational model) you have thus to print: nll_attach_y, nll_attach_event, nll_attach

**Comment by NicolasGensollen** [comment]
This piece of logic seems to belong to the model rather than here, in the algorithm. The model would expose a public method to get the loss which would internally handle the different cases where the model can have one or several observation models. Maybe it even exists already, we should take a closer look. WDYT ?

Also, sorry if this is obvious, but why would the loss be zero if the first observation model isn't a Gaussian model ?

**Comment by NicolasGensollen** [comment]
Could you check whether this is still used. If not, I agree that we should delete it.

**Comment by NicolasGensollen** [comment]
```suggestion:-0+0
        for i, parameter_name in enumerate(params_to_plot):
```

**Comment by NicolasGensollen** [comment]
Same kind of comment, with Pathlib this would become:

```python
import_path = self.path_save_model_parameters_convergence / f"{parameter_name}.csv"
```

**Comment by NicolasGensollen** [comment]
```suggestion:-0+0
            true_values_patient = data.get_values_patient(i).cpu().detach().numpy()
```

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Jan 24, 2025, 10:15

Commented on [leaspy/io/logs/fit_output_manager.py line 175](https://github.com/aramis-lab/leaspy/compare/1d109789625fcaaf2eb372af613c453381e9761f..cbf718c618e66e642a6d0b019385f23559872082#diff-e3aa1651244769303da0f561fed27e42R175)



changed this line in [version 4 of the diff](/icm-institute/aramislab/leaspy/-/merge_requests/179/diffs?diff_id=1243836426&start_sha=cbf718c618e66e642a6d0b019385f23559872082#e6199329d82f2e81a42750b961f4bd6acc56f7da_175_175)

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Jan 24, 2025, 10:15

Commented on [leaspy/io/logs/fit_output_manager.py line 230](https://github.com/aramis-lab/leaspy/compare/1d109789625fcaaf2eb372af613c453381e9761f..cbf718c618e66e642a6d0b019385f23559872082#diff-e3aa1651244769303da0f561fed27e42R230)



changed this line in [version 4 of the diff](/icm-institute/aramislab/leaspy/-/merge_requests/179/diffs?diff_id=1243836426&start_sha=cbf718c618e66e642a6d0b019385f23559872082#e6199329d82f2e81a42750b961f4bd6acc56f7da_230_230)

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Jan 24, 2025, 10:15

Commented on [leaspy/io/logs/fit_output_manager.py line 231](https://github.com/aramis-lab/leaspy/compare/1d109789625fcaaf2eb372af613c453381e9761f..cbf718c618e66e642a6d0b019385f23559872082#diff-e3aa1651244769303da0f561fed27e42R231)



changed this line in [version 4 of the diff](/icm-institute/aramislab/leaspy/-/merge_requests/179/diffs?diff_id=1243836426&start_sha=cbf718c618e66e642a6d0b019385f23559872082#e6199329d82f2e81a42750b961f4bd6acc56f7da_231_230)

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Jan 24, 2025, 10:15

Commented on [leaspy/io/logs/fit_output_manager.py line 234](https://github.com/aramis-lab/leaspy/compare/1d109789625fcaaf2eb372af613c453381e9761f..cbf718c618e66e642a6d0b019385f23559872082#diff-e3aa1651244769303da0f561fed27e42R234)



changed this line in [version 4 of the diff](/icm-institute/aramislab/leaspy/-/merge_requests/179/diffs?diff_id=1243836426&start_sha=cbf718c618e66e642a6d0b019385f23559872082#e6199329d82f2e81a42750b961f4bd6acc56f7da_234_234)

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Jan 24, 2025, 10:15

added 1 commit

<ul><li>c27e039c - apply suggestions</li></ul>

[Compare with previous version](/icm-institute/aramislab/leaspy/-/merge_requests/179/diffs?diff_id=1243836426&start_sha=cbf718c618e66e642a6d0b019385f23559872082)

**Comment by NicolasGensollen** [comment]
In GitLab by @JulietteOrtholand on Jan 24, 2025, 11:53

Hey,

Thank you for your MR, just small comments on the objectives of the MR:

[Point 1] For the convergence parameters that are plotted. I think we need to discuss that, as they are of different nature. For example, you want to minimise the nll_attach, but want the parameters of the model (such as g, v0) to be stable. I think we need to order them in the PDF, or maybe make several, let’s discuss that next Tuesday, maybe?

[Point 2] Is this issue solve or still pending after our discussions?
 
[Point 6] Did you solve your issue after we discussed `compute_individual_tensorized`?

My next step would be to test it, thanks for sharing the tool example I’ll try to check that ASAP.

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Jan 24, 2025, 14:49

Commented on [leaspy/algo/fit/abstract_fit_algo.py line 126](https://github.com/aramis-lab/leaspy/compare/1d109789625fcaaf2eb372af613c453381e9761f..c27e039c17d1a2a617552ded838879e27c647edd#diff-39f008a75a12b6fbdb801e2fd2a6f058R126)



changed this line in [version 5 of the diff](/icm-institute/aramislab/leaspy/-/merge_requests/179/diffs?diff_id=1244196474&start_sha=c27e039c17d1a2a617552ded838879e27c647edd#97a8dcc83618236577918c38574acb23d0c54ae4_126_126)

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Jan 24, 2025, 14:49

Commented on [leaspy/algo/fit/abstract_fit_algo.py line 219](https://github.com/aramis-lab/leaspy/compare/1d109789625fcaaf2eb372af613c453381e9761f..c27e039c17d1a2a617552ded838879e27c647edd#diff-39f008a75a12b6fbdb801e2fd2a6f058R219)



changed this line in [version 5 of the diff](/icm-institute/aramislab/leaspy/-/merge_requests/179/diffs?diff_id=1244196474&start_sha=c27e039c17d1a2a617552ded838879e27c647edd#97a8dcc83618236577918c38574acb23d0c54ae4_219_219)

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Jan 24, 2025, 14:49

Commented on [leaspy/io/logs/fit_output_manager.py line 55](https://github.com/aramis-lab/leaspy/compare/1d109789625fcaaf2eb372af613c453381e9761f..c27e039c17d1a2a617552ded838879e27c647edd#diff-e3aa1651244769303da0f561fed27e42R55)



changed this line in [version 5 of the diff](/icm-institute/aramislab/leaspy/-/merge_requests/179/diffs?diff_id=1244196474&start_sha=c27e039c17d1a2a617552ded838879e27c647edd#e6199329d82f2e81a42750b961f4bd6acc56f7da_55_56)

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Jan 24, 2025, 14:49

Commented on [leaspy/io/logs/fit_output_manager.py line 169](https://github.com/aramis-lab/leaspy/compare/1d109789625fcaaf2eb372af613c453381e9761f..c27e039c17d1a2a617552ded838879e27c647edd#diff-e3aa1651244769303da0f561fed27e42R169)



changed this line in [version 5 of the diff](/icm-institute/aramislab/leaspy/-/merge_requests/179/diffs?diff_id=1244196474&start_sha=c27e039c17d1a2a617552ded838879e27c647edd#e6199329d82f2e81a42750b961f4bd6acc56f7da_169_168)

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Jan 24, 2025, 14:49

Commented on [leaspy/io/logs/fit_output_manager.py line 176](https://github.com/aramis-lab/leaspy/compare/1d109789625fcaaf2eb372af613c453381e9761f..c27e039c17d1a2a617552ded838879e27c647edd#diff-e3aa1651244769303da0f561fed27e42R176)



changed this line in [version 5 of the diff](/icm-institute/aramislab/leaspy/-/merge_requests/179/diffs?diff_id=1244196474&start_sha=c27e039c17d1a2a617552ded838879e27c647edd#e6199329d82f2e81a42750b961f4bd6acc56f7da_176_175)

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Jan 24, 2025, 14:49

Commented on [leaspy/io/logs/fit_output_manager.py line 220](https://github.com/aramis-lab/leaspy/compare/1d109789625fcaaf2eb372af613c453381e9761f..c27e039c17d1a2a617552ded838879e27c647edd#diff-e3aa1651244769303da0f561fed27e42R220)



changed this line in [version 5 of the diff](/icm-institute/aramislab/leaspy/-/merge_requests/179/diffs?diff_id=1244196474&start_sha=c27e039c17d1a2a617552ded838879e27c647edd#e6199329d82f2e81a42750b961f4bd6acc56f7da_220_216)

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Jan 24, 2025, 14:49

added 1 commit

<ul><li>90652efc - follow review</li></ul>

[Compare with previous version](/icm-institute/aramislab/leaspy/-/merge_requests/179/diffs?diff_id=1244196474&start_sha=c27e039c17d1a2a617552ded838879e27c647edd)

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Jan 24, 2025, 14:50

Commented on [leaspy/algo/fit/abstract_fit_algo.py line 126](https://github.com/aramis-lab/leaspy/compare/1d109789625fcaaf2eb372af613c453381e9761f..c27e039c17d1a2a617552ded838879e27c647edd#diff-39f008a75a12b6fbdb801e2fd2a6f058R126)



I removed that part of the comment

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Jan 24, 2025, 14:53

Commented on [leaspy/algo/fit/abstract_fit_algo.py line 219](https://github.com/aramis-lab/leaspy/compare/1d109789625fcaaf2eb372af613c453381e9761f..c27e039c17d1a2a617552ded838879e27c647edd#diff-39f008a75a12b6fbdb801e2fd2a6f058R219)



Removed it

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Jan 24, 2025, 15:05

Commented on [leaspy/algo/fit/abstract_fit_algo.py line 135](https://github.com/aramis-lab/leaspy/compare/1d109789625fcaaf2eb372af613c453381e9761f..b8b2780792d8f73b7ceaa01770aef65739652939#diff-39f008a75a12b6fbdb801e2fd2a6f058R135)



This is the only part in this MR which is still unsolved from my side.  You can also read the discussion above with Juliette. [Here](https://gitlab.com/icm-institute/aramislab/leaspy/-/blob/v2/leaspy/algo/abstract_algo.py?ref_type=heads#L174), this loss is only used to print some meaningful values for the user. Noise_std is not a loss for gaussian or for other observational models. So maybe, we need to change the term `loss` here and we can say `values_to_print` which could be different for each observational model. Or, we can print last loss which is `nll_attach` for every observational models with some modifications for joint model.

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Jan 24, 2025, 15:05

Commented on [leaspy/io/logs/fit_output_manager.py line 55](https://github.com/aramis-lab/leaspy/compare/1d109789625fcaaf2eb372af613c453381e9761f..c27e039c17d1a2a617552ded838879e27c647edd#diff-e3aa1651244769303da0f561fed27e42R55)



I use Pathlib now. is it better?

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Jan 24, 2025, 15:05

Commented on [leaspy/io/logs/fit_output_manager.py line 169](https://github.com/aramis-lab/leaspy/compare/1d109789625fcaaf2eb372af613c453381e9761f..c27e039c17d1a2a617552ded838879e27c647edd#diff-e3aa1651244769303da0f561fed27e42R169)



Thanks, I changed it

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Jan 24, 2025, 15:06

Commented on [leaspy/io/logs/fit_output_manager.py line 163](https://github.com/aramis-lab/leaspy/compare/aeaa8db0eb21fd4866a1ffe7a818d3ebc3306887..a46ed3fabf0d9f3a9eccaf0f71204ad283686fbb#diff-e3aa1651244769303da0f561fed27e42R163)



I agree

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Jan 24, 2025, 15:06

Commented on [leaspy/io/logs/fit_output_manager.py line 176](https://github.com/aramis-lab/leaspy/compare/1d109789625fcaaf2eb372af613c453381e9761f..c27e039c17d1a2a617552ded838879e27c647edd#diff-e3aa1651244769303da0f561fed27e42R176)



thanks, I did that

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Jan 24, 2025, 15:07

Commented on [leaspy/io/logs/fit_output_manager.py line 199](https://github.com/aramis-lab/leaspy/compare/aeaa8db0eb21fd4866a1ffe7a818d3ebc3306887..a46ed3fabf0d9f3a9eccaf0f71204ad283686fbb#diff-e3aa1651244769303da0f561fed27e42R199)



Not my choice, it was like this before. If there are several patients, figures will be very complicated

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Jan 24, 2025, 15:07

Commented on [leaspy/io/logs/fit_output_manager.py line 220](https://github.com/aramis-lab/leaspy/compare/1d109789625fcaaf2eb372af613c453381e9761f..c27e039c17d1a2a617552ded838879e27c647edd#diff-e3aa1651244769303da0f561fed27e42R220)



done

**Comment by NicolasGensollen** [comment]
```suggestion:-0+0
            self.path_plot_convergence_model_parameters = self.path_plot / "convergence_parameters.pdf"
```

**Comment by NicolasGensollen** [comment]
`append` is a list method.

```suggestion:-0+0
            to_skip.add("deltas")
```

**Comment by NicolasGensollen** [comment]
Since you're using this variable only once in the next line, I would remove this and modify next line to be:

```python
f_convergence = pd.read_csv(
    self.path_save_model_parameters_convergence / f"{parameter_name}.csv",
    index_col=0,
    header=None,
)
```

**Comment by NicolasGensollen** [comment]
Ideally, I think this should be configurable by the user. 5 could be a default, set for readability concerns as you said, but definitely not a hardcoded limit.

**Comment by NicolasGensollen** [comment]
Looking at the plots, it feels like the users might want a legend specifying the patient ID for each curve. In this case, the title could be something a little bit more informative like "Feature trajectory for patients 1 to 5".
WDYT ?

**Comment by NicolasGensollen** [comment]
Great ! Closing this for now then !

**Comment by NicolasGensollen** [comment]
I think we need to clarify this indeed. I'll take a closer look at the discussion you linked, but it'd be helpful (to me at least) to know exactly what we are computing for the different cases (one or several observation models, types of these models...).

Once we have a better understanding of this, we'd be able to come up with a meaningful name (`values_to_print` is not very informative of what is computed I think). Maybe `goodness_of_fit` of something like that ? :thinking: 

Let's talk about this tomorrow.

**Comment by NicolasGensollen** [comment]
Thanks for taking my suggestions into account @caglayantuna ! 

I tried the code you provided and it worked as expected.

I made some additional comments and suggestions.

I think there are two important points we need to discuss tomorrow:

- the look of the PDF reports as mentioned by @JulietteOrtholand 
- what is exactly the observational model `loss`

**Comment by NicolasGensollen** [comment]
Thanks !

**Comment by NicolasGensollen** [comment]
Yes, thanks !

**Comment by NicolasGensollen** [comment]
Thanks

**Comment by NicolasGensollen** [comment]
Thanks

**Comment by NicolasGensollen** [comment]
If I understand correctly, the model should have:

- an attribute (or a method) returning the full loss (seems to be `nll_attached` atm)
- a method returning the name and loss of each observation model (maybe it already exists, otherwise something like `get_observation_models_loss`)

Something that would look more or less like:

```python
def get_observation_models_loss(self) -> dict[VarName, float]
```

In the example of a Joint model this would give something like that:

```python
>>> model.nll_attached
0.3
>>> model.get_observation_models_loss()
{
    "nll_attach_y": 0.1,
    "nll_attach_event": 0.2,
}
```

Then the algorithm could use those two methods to display information in the desired way (table or whatever).

WDYT ?

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Jan 28, 2025, 14:36

Commented on [leaspy/io/logs/fit_output_manager.py line 56](https://github.com/aramis-lab/leaspy/compare/1d109789625fcaaf2eb372af613c453381e9761f..90652efcdfd6d9fbb2c1d3dcbdf9bb5da4930ce8#diff-e3aa1651244769303da0f561fed27e42R56)



changed this line in [version 6 of the diff](/icm-institute/aramislab/leaspy/-/merge_requests/179/diffs?diff_id=1246998703&start_sha=90652efcdfd6d9fbb2c1d3dcbdf9bb5da4930ce8#e6199329d82f2e81a42750b961f4bd6acc56f7da_56_56)

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Jan 28, 2025, 14:36

added 1 commit

<ul><li>b8b27807 - apply suggestions</li></ul>

[Compare with previous version](/icm-institute/aramislab/leaspy/-/merge_requests/179/diffs?diff_id=1246998703&start_sha=90652efcdfd6d9fbb2c1d3dcbdf9bb5da4930ce8)

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Jan 29, 2025, 16:59

Commented on [leaspy/algo/fit/abstract_fit_algo.py line 135](https://github.com/aramis-lab/leaspy/compare/1d109789625fcaaf2eb372af613c453381e9761f..b8b2780792d8f73b7ceaa01770aef65739652939#diff-39f008a75a12b6fbdb801e2fd2a6f058R135)



changed this line in [version 7 of the diff](/icm-institute/aramislab/leaspy/-/merge_requests/179/diffs?diff_id=1248638312&start_sha=b8b2780792d8f73b7ceaa01770aef65739652939#97a8dcc83618236577918c38574acb23d0c54ae4_135_134)

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Jan 29, 2025, 16:59

added 1 commit

<ul><li>023fd5aa - update MR</li></ul>

[Compare with previous version](/icm-institute/aramislab/leaspy/-/merge_requests/179/diffs?diff_id=1248638312&start_sha=b8b2780792d8f73b7ceaa01770aef65739652939)

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Jan 30, 2025, 11:57

Commented on [leaspy/io/logs/fit_output_manager.py line 175](https://github.com/aramis-lab/leaspy/compare/1d109789625fcaaf2eb372af613c453381e9761f..b8b2780792d8f73b7ceaa01770aef65739652939#diff-e3aa1651244769303da0f561fed27e42R175)



done

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Jan 30, 2025, 13:18

Commented on [leaspy/io/logs/fit_output_manager.py line 199](https://github.com/aramis-lab/leaspy/compare/aeaa8db0eb21fd4866a1ffe7a818d3ebc3306887..a46ed3fabf0d9f3a9eccaf0f71204ad283686fbb#diff-e3aa1651244769303da0f561fed27e42R199)



I added `nb_of_patients_to_plot` for user

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Jan 30, 2025, 13:28

I did some changes in this MR according to your comments @NicolasGensollen . However, about legend of `plot_patient_reconstruction` and pdf creating without csv files are missing. I would be happy to know your suggestions.

I also did some small changes such as `console_print_periodicity` to `print_periodicity`, `calibration` to `fit` and I removed `n_realizations`. In addition, I removed `loss_repr` function which is not useful anymore.

**Comment by NicolasGensollen** [comment]
In GitLab by @JulietteOrtholand on Jan 30, 2025, 13:49

Commented on [leaspy/algo/fit/abstract_fit_algo.py line 135](https://github.com/aramis-lab/leaspy/compare/1d109789625fcaaf2eb372af613c453381e9761f..b8b2780792d8f73b7ceaa01770aef65739652939#diff-39f008a75a12b6fbdb801e2fd2a6f058R135)



What about saving the logs in any case ?

**Comment by NicolasGensollen** [comment]
I think this would make sense. At the moment, when you instantiate the `AlgorithmSettings` class, the attribute `logs` is None until you call the `set_logs` method which instantiate the `OutputsSettings` class with either default values, or values read from a file.

I think it'd make sense to have the `logs` attribute instantiated in the `AlgorithmSettings` constructor with default values and the `set_logs` method could then be used to replace the settings with other settings, either as method parameters or as a JSON file.

In this case, all algorithm will have logging enabled by default with default parameter values.

WDYT ?

**Comment by NicolasGensollen** [comment]
Thanks !

**Comment by NicolasGensollen** [comment]
Following our discussions from Tuesday and the decision to remove the "loss" and let the user decide visually on the convergence, I'm closing the thread.

**Comment by NicolasGensollen** [comment]
Thanks !

**Comment by NicolasGensollen** [comment]
It might be a bit complicated to do and look messy with legends. Maybe a text box with the patient ID close to the corresponding curve would be easier to do and look nicer ?

If that's too complicated, we can forget about this.

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Jan 31, 2025, 16:27

added 1 commit

<ul><li>a766b920 - update</li></ul>

[Compare with previous version](/icm-institute/aramislab/leaspy/-/merge_requests/179/diffs?diff_id=1251297506&start_sha=023fd5aa3f713ebe08498997bf65044814507ff6)

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Jan 31, 2025, 16:31

Commented on [leaspy/algo/fit/abstract_fit_algo.py line 135](https://github.com/aramis-lab/leaspy/compare/1d109789625fcaaf2eb372af613c453381e9761f..b8b2780792d8f73b7ceaa01770aef65739652939#diff-39f008a75a12b6fbdb801e2fd2a6f058R135)



I changed some lines in `algo_factory` to handle this. I couldn't find a proper way in `algorithmsettings` or in `abstract_algo`. With this change, only fit algorithm (not personalize!) creates _outputs folder and saves the results with default values without user input for logs.

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Jan 31, 2025, 16:33

Commented on [leaspy/io/logs/fit_output_manager.py line 216](https://github.com/aramis-lab/leaspy/compare/1d109789625fcaaf2eb372af613c453381e9761f..b8b2780792d8f73b7ceaa01770aef65739652939#diff-e3aa1651244769303da0f561fed27e42R216)



You can check new figure with last change. Now, we have patient ID's near to curves. is it ok?undefined

**Comment by NicolasGensollen** [comment]
Isn't this logic equivalent to:

```python
if settings.logs is None and settings.algo_class.family == "fit":
    settings.set_logs()
algorithm.set_output_manager(settings.logs)
```

??

**Comment by NicolasGensollen** [comment]
I think it looks great ! Thanks for implementing it.

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Jan 31, 2025, 16:55

Commented on [leaspy/algo/algo_factory.py line 116](https://github.com/aramis-lab/leaspy/compare/1d109789625fcaaf2eb372af613c453381e9761f..a766b920e5081cdeadd92804aeaa26ced8287137#diff-8d5cbc6570e10e406eebc9a1e0e8b794R116)



changed this line in [version 9 of the diff](/icm-institute/aramislab/leaspy/-/merge_requests/179/diffs?diff_id=1251336579&start_sha=a766b920e5081cdeadd92804aeaa26ced8287137#91c7729e09ae68775b85e819e328f053bee497fd_116_108)

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Jan 31, 2025, 16:55

added 1 commit

<ul><li>9595fdf3 - shorten conditions</li></ul>

[Compare with previous version](/icm-institute/aramislab/leaspy/-/merge_requests/179/diffs?diff_id=1251336579&start_sha=a766b920e5081cdeadd92804aeaa26ced8287137)

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Jan 31, 2025, 16:55

Commented on [leaspy/algo/algo_factory.py line 116](https://github.com/aramis-lab/leaspy/compare/1d109789625fcaaf2eb372af613c453381e9761f..a766b920e5081cdeadd92804aeaa26ced8287137#diff-8d5cbc6570e10e406eebc9a1e0e8b794R116)



yes of course, thanks, changed it

**Comment by NicolasGensollen** [comment]
Ok, thanks. Let's keep it like this for now and move on. I'm closing the thread.

**Comment by NicolasGensollen** [comment]
resolved all threads

**Comment by NicolasGensollen** [comment]
@caglayantuna Looking at the CI, it seems like there are a lot of broken tests.

Could you first make sure your branch is up to date with V2 (and rebase if not), and look at those failing tests ?

**Comment by NicolasGensollen** [comment]
In GitLab by @caglayantuna on Feb 5, 2025, 16:05

added 16 commits

<ul><li>9595fdf3...aeaa8db0 - 6 commits from branch <code>v2</code></li><li>91348c93 - basic version of FOM for v2</li><li>d800a97a - fix_tests</li><li>59498199 - adds plot patient rec and first attemp for generalized loss</li><li>0d230f8e - apply suggestions</li><li>cd96a091 - follow review</li><li>aa5e2785 - apply suggestions</li><li>4d7cd9c3 - update MR</li><li>7e6e168c - update</li><li>e28d0cb3 - shorten conditions</li><li>a46ed3fa - fix loss issue</li></ul>

[Compare with previous version](/icm-institute/aramislab/leaspy/-/merge_requests/179/diffs?diff_id=1255572066&start_sha=9595fdf34fe751a9895ea96bac6046cdb1ed4b1e)
